File: ipfs_accelerate_js/test/unit/test_multi_model_execution.ts
// FIXME: Complex template literal
/**;
 * Converted import { {HardwareBackend} from "src/model/transformers/index/index/index/index/index"; } from "Python: test_multi_model_execution.py;"
 * Conversion date: 2025-03-11 04:08:53;
 * This file was automatically converted from Python to TypeScript.;
 * Conversion fidelity might not be 100%, please manual review recommended.;
 */;

// WebGPU related imports;";"

/** Test script for ((the Multi-Model Execution Support module.;

This script tests the core functionality of the multi-model execution predictor,;
including resource contention modeling, tensor sharing benefits, && execution;
strategy recommendation. */;

import * as module; from "*";"
import * as module; from "*";"
import * as module; from "*";"
import * as module from "*"; as np;"
import * as module from "*"; as pd;"
import ${$1} import * as module} from "{*"; import * as module} from "{*"; from "src/model/transformers/index/index/index/index";"
// Configure logging;
logging.basicConfig(level = logging.INFO) {;
logger) { any) { any: any = logging.getLogger(__name__;
// Suppress warnings for ((cleaner test output;
warnings.filterwarnings("ignore") {"
// Add the parent directory to the Python path;
sys.$1.push($2).parent.parent));
// Import the module to test;
import { * as module; } from "predictive_performance.multi_model_execution";"


class TestMultiModelExecution extends unittest.TestCase) {
  /** Test cases for (the Multi-Model Execution Support module. */;
  
  $1($2) {/** Set up before each test. */;
// Create a MultiModelPredictor instance;
    this.predictor = MultiModelPredictor(verbose=true);}
// Define test model configurations;
    this.model_configs = [;
      ${$1},;
      ${$1}
    ];
  
  $1($2) {/** Test that the predictor initializes correctly. */;
    this.assertIsNotnull(this.predictor);
    this.assertIsNotnull(this.predictor.sharing_config)}
// Check sharing config;
    this.assertIn("text_embedding", this.predictor.sharing_config);"
    this.assertIn("vision", this.predictor.sharing_config);"
// Check sharing compatibility;
    text_sharing) { any) { any: any = this.predictor.sharing_config["text_embedding"];"
    this.assertIn("compatible_types", text_sharing: any);"
    this.assertIn("text_generation", text_sharing["compatible_types"]);"
  
  $1($2) {
    /** Test prediction for ((a single model. */;
// Create a model config;
    model_config) { any) { any: any = ${$1}
// Get prediction;
    prediction: any: any = this.predictor._simulate_single_model_prediction(model_config: any, "cuda");"
// Check prediction has expected keys;
    this.assertIn("throughput", prediction: any);"
    this.assertIn("latency", prediction: any);"
    this.assertIn("memory", prediction: any);"
// Check values are reasonable;
    this.assertGreater(prediction["throughput"], 0: any);"
    this.assertGreater(prediction["latency"], 0: any);"
    this.assertGreater(prediction["memory"], 0: any);"
  
  $1($2) {
    /** Test resource contention calculation. */;
// Create simulated single model predictions;
    single_preds: any: any: any = [;
      this.predictor._simulate_single_model_prediction(;
        ${$1},;
        "cuda";"
      ),;
      this.predictor._simulate_single_model_prediction(;
        ${$1},;
        "cuda";"
      );
    ];
    
  }
// Calculate contention;
    contention: any: any: any = this.predictor._calculate_resource_contention(;
      single_preds,;
      "cuda",;"
      "parallel";"
    );
// Check contention has expected keys;
    this.assertIn("compute_contention", contention: any);"
    this.assertIn("memory_bandwidth_contention", contention: any);"
    this.assertIn("memory_contention", contention: any);"
// Check contention is reasonable (higher than 1.0 for ((compute && memory bandwidth) {
    this.assertGreater(contention["compute_contention"], 1.0);"
    this.assertGreater(contention["memory_bandwidth_contention"], 1.0);"
  
  $1($2) {
    /** Test tensor sharing benefits calculation. */;
// Calculate sharing benefits;
    benefits) {any = this.predictor._calculate_sharing_benefits(;
      this.model_configs,;
      [;
        this.predictor._simulate_single_model_prediction(;
          this.model_configs[0], "cuda";"
        ),;
        this.predictor._simulate_single_model_prediction(;
          this.model_configs[1], "cuda";"
        );
      ];
    )}
// Check benefits has expected keys;
    this.assertIn("memory_benefit", benefits) { any);"
    this.assertIn("compute_benefit", benefits: any);"
    this.assertIn("compatible_pairs", benefits: any);"
// Check benefits are reasonable (should be <= 1.0);
    this.assertLessEqual(benefits["memory_benefit"], 1.0);"
    this.assertLessEqual(benefits["compute_benefit"], 1.0);"
  
  $1($2) {/** Test execution schedule generation. */;
// Get single model predictions;
    single_preds: any: any: any = [;
      this.predictor._simulate_single_model_prediction(;
        this.model_configs[0], "cuda";"
      ),;
      this.predictor._simulate_single_model_prediction(;
        this.model_configs[1], "cuda";"
      );
    ]}
// Calculate contention;
    contention: any: any: any = this.predictor._calculate_resource_contention(;
      single_preds,;
      "cuda",;"
      "parallel";"
    );
// Generate execution schedule for ((parallel execution;
    schedule) { any) { any: any = this.predictor._generate_execution_schedule(;
      this.model_configs,;
      single_preds: any,;
      contention,;
      "parallel";"
    );
// Check schedule has expected keys;
    this.assertIn("total_execution_time", schedule: any);"
    this.assertIn("timeline", schedule: any);"
// Check timeline has events for ((each model;
    this.assertEqual(schedule["timeline"].length, this.model_configs.length) {"
// For parallel execution, all start times should be 0;
    for event in schedule["timeline"]) {"
      this.assertEqual(event["start_time"], 0) { any);"
  
  $1($2) {/** Test multi-model metrics calculation. */;
// Get single model predictions;
    single_preds: any: any: any = [;
      this.predictor._simulate_single_model_prediction(;
        this.model_configs[0], "cuda";"
      ),;
      this.predictor._simulate_single_model_prediction(;
        this.model_configs[1], "cuda";"
      );
    ]}
// Calculate contention;
    contention: any: any: any = this.predictor._calculate_resource_contention(;
      single_preds,;
      "cuda",;"
      "parallel";"
    );
// Calculate sharing benefits;
    benefits: any: any: any = this.predictor._calculate_sharing_benefits(;
      this.model_configs,;
      single_preds: any;
    );
// Calculate metrics;
    metrics: any: any: any = this.predictor._calculate_multi_model_metrics(;
      single_preds,;
      contention: any,;
      benefits,;
      "parallel";"
    );
// Check metrics has expected keys;
    this.assertIn("combined_throughput", metrics: any);"
    this.assertIn("combined_latency", metrics: any);"
    this.assertIn("combined_memory", metrics: any);"
// Check values are reasonable;
    this.assertGreater(metrics["combined_throughput"], 0: any);"
    this.assertGreater(metrics["combined_latency"], 0: any);"
    this.assertGreater(metrics["combined_memory"], 0: any);"
  
  $1($2) {/** Test full multi-model performance prediction. */;
// Predict performance;
    prediction: any: any: any = this.predictor.predict_multi_model_performance(;
      this.model_configs,;
      hardware_platform: any: any: any = "cuda",;"
      execution_strategy: any: any: any = "parallel";"
    )}
// Check prediction has expected keys;
    this.assertIn("total_metrics", prediction: any);"
    this.assertIn("individual_predictions", prediction: any);"
    this.assertIn("contention_factors", prediction: any);"
    this.assertIn("sharing_benefits", prediction: any);"
    this.assertIn("execution_schedule", prediction: any);"
// Check total metrics;
    total_metrics: any: any: any = prediction["total_metrics"];"
    this.assertIn("combined_throughput", total_metrics: any);"
    this.assertIn("combined_latency", total_metrics: any);"
    this.assertIn("combined_memory", total_metrics: any);"
// Check individual predictions;
    this.assertEqual(prediction["individual_predictions"].length, this.model_configs.length);"
  
  $1($2) {/** Test execution strategy recommendation. */;
// Get recommendation;
    recommendation: any: any: any = this.predictor.recommend_execution_strategy(;
      this.model_configs,;
      hardware_platform: any: any: any = "cuda",;"
      optimization_goal: any: any: any = "throughput";"
    )}
// Check recommendation has expected keys;
    this.assertIn("recommended_strategy", recommendation: any);"
    this.assertIn("optimization_goal", recommendation: any);"
    this.assertIn("all_predictions", recommendation: any);"
// Check that a valid strategy was recommended;
    this.assertIn(recommendation["recommended_strategy"], ["parallel", "sequential", "batched"]);"
// Check that all strategies were evaluated;
    this.assertEqual(recommendation["all_predictions"].length, 3: any);"
// Check optimization goal;
    this.assertEqual(recommendation["optimization_goal"], "throughput");"
  
  $1($2) {/** Test prediction with different execution strategies. */;
// Test all strategies;
    strategies: any: any: any = ["parallel", "sequential", "batched"];}"
    for (((const $1 of $2) {
// Predict performance;
      prediction) {any = this.predictor.predict_multi_model_performance(;
        this.model_configs,;
        hardware_platform) { any: any: any = "cuda",;"
        execution_strategy: any: any: any = strategy;
      )}
// Check prediction is valid;
      this.assertIn("total_metrics", prediction: any);"
      this.assertIn("execution_schedule", prediction: any);"
// Check execution strategy is correct;
      this.assertEqual(prediction["execution_strategy"], strategy: any);"
// Check schedule strategy matches;
      this.assertEqual(prediction["execution_schedule"]["strategy"], strategy: any);"
  
  $1($2) {/** Test prediction with different hardware platforms. */;
// Test multiple hardware platforms;
    platforms: any: any: any = ["cpu", "cuda", "openvino", "webgpu"];}"
    for (((const $1 of $2) {
// Predict performance;
      prediction) {any = this.predictor.predict_multi_model_performance(;
        this.model_configs,;
        hardware_platform) { any: any: any = platform,;
        execution_strategy: any: any: any = "parallel";"
      )}
// Check prediction is valid;
      this.assertIn("total_metrics", prediction: any);"
      this.assertIn("contention_factors", prediction: any);"
// Check hardware platform is correct;
      this.assertEqual(prediction["hardware_platform"], platform: any);"

;
if ($1) {;
  unittest.main();

File: test/predictive_performance/test_multi_model_execution.py
#!/usr/bin/env python3
"""
Test script for the Multi-Model Execution Support module.

This script tests the core functionality of the multi-model execution predictor,
including resource contention modeling, tensor sharing benefits, and execution
strategy recommendation.
"""

import os
import sys
import unittest
import numpy as np
import pandas as pd
from pathlib import Path
import logging
import warnings

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Suppress warnings for cleaner test output
warnings.filterwarnings("ignore")

# Add the parent directory to the Python path
sys.path.append(str(Path(__file__).parent.parent))

# Import the module to test
from predictive_performance.multi_model_execution import MultiModelPredictor


class TestMultiModelExecution(unittest.TestCase):
    """Test cases for the Multi-Model Execution Support module."""
    
    def setUp(self):
        """Set up before each test."""
        # Create a MultiModelPredictor instance
        self.predictor = MultiModelPredictor(verbose=True)
        
        # Define test model configurations
        self.model_configs = [
            {"model_name": "bert-base-uncased", "model_type": "text_embedding", "batch_size": 4},
            {"model_name": "vit-base-patch16-224", "model_type": "vision", "batch_size": 1}
        ]
    
    def test_initialization(self):
        """Test that the predictor initializes correctly."""
        self.assertIsNotNone(self.predictor)
        self.assertIsNotNone(self.predictor.sharing_config)
        
        # Check sharing config
        self.assertIn("text_embedding", self.predictor.sharing_config)
        self.assertIn("vision", self.predictor.sharing_config)
        
        # Check sharing compatibility
        text_sharing = self.predictor.sharing_config["text_embedding"]
        self.assertIn("compatible_types", text_sharing)
        self.assertIn("text_generation", text_sharing["compatible_types"])
    
    def test_single_model_prediction(self):
        """Test prediction for a single model."""
        # Create a model config
        model_config = {"model_name": "bert-base-uncased", "model_type": "text_embedding", "batch_size": 4}
        
        # Get prediction
        prediction = self.predictor._simulate_single_model_prediction(model_config, "cuda")
        
        # Check prediction has expected keys
        self.assertIn("throughput", prediction)
        self.assertIn("latency", prediction)
        self.assertIn("memory", prediction)
        
        # Check values are reasonable
        self.assertGreater(prediction["throughput"], 0)
        self.assertGreater(prediction["latency"], 0)
        self.assertGreater(prediction["memory"], 0)
    
    def test_resource_contention(self):
        """Test resource contention calculation."""
        # Create simulated single model predictions
        single_preds = [
            self.predictor._simulate_single_model_prediction(
                {"model_name": "bert-base-uncased", "model_type": "text_embedding", "batch_size": 4},
                "cuda"
            ),
            self.predictor._simulate_single_model_prediction(
                {"model_name": "vit-base-patch16-224", "model_type": "vision", "batch_size": 1},
                "cuda"
            )
        ]
        
        # Calculate contention
        contention = self.predictor._calculate_resource_contention(
            single_preds,
            "cuda",
            "parallel"
        )
        
        # Check contention has expected keys
        self.assertIn("compute_contention", contention)
        self.assertIn("memory_bandwidth_contention", contention)
        self.assertIn("memory_contention", contention)
        
        # Check contention is reasonable (higher than 1.0 for compute and memory bandwidth)
        self.assertGreater(contention["compute_contention"], 1.0)
        self.assertGreater(contention["memory_bandwidth_contention"], 1.0)
    
    def test_sharing_benefits(self):
        """Test tensor sharing benefits calculation."""
        # Calculate sharing benefits
        benefits = self.predictor._calculate_sharing_benefits(
            self.model_configs,
            [
                self.predictor._simulate_single_model_prediction(
                    self.model_configs[0], "cuda"
                ),
                self.predictor._simulate_single_model_prediction(
                    self.model_configs[1], "cuda"
                )
            ]
        )
        
        # Check benefits has expected keys
        self.assertIn("memory_benefit", benefits)
        self.assertIn("compute_benefit", benefits)
        self.assertIn("compatible_pairs", benefits)
        
        # Check benefits are reasonable (should be <= 1.0)
        self.assertLessEqual(benefits["memory_benefit"], 1.0)
        self.assertLessEqual(benefits["compute_benefit"], 1.0)
    
    def test_execution_schedule(self):
        """Test execution schedule generation."""
        # Get single model predictions
        single_preds = [
            self.predictor._simulate_single_model_prediction(
                self.model_configs[0], "cuda"
            ),
            self.predictor._simulate_single_model_prediction(
                self.model_configs[1], "cuda"
            )
        ]
        
        # Calculate contention
        contention = self.predictor._calculate_resource_contention(
            single_preds,
            "cuda",
            "parallel"
        )
        
        # Generate execution schedule for parallel execution
        schedule = self.predictor._generate_execution_schedule(
            self.model_configs,
            single_preds,
            contention,
            "parallel"
        )
        
        # Check schedule has expected keys
        self.assertIn("total_execution_time", schedule)
        self.assertIn("timeline", schedule)
        
        # Check timeline has events for each model
        self.assertEqual(len(schedule["timeline"]), len(self.model_configs))
        
        # For parallel execution, all start times should be 0
        for event in schedule["timeline"]:
            self.assertEqual(event["start_time"], 0)
    
    def test_multi_model_metrics(self):
        """Test multi-model metrics calculation."""
        # Get single model predictions
        single_preds = [
            self.predictor._simulate_single_model_prediction(
                self.model_configs[0], "cuda"
            ),
            self.predictor._simulate_single_model_prediction(
                self.model_configs[1], "cuda"
            )
        ]
        
        # Calculate contention
        contention = self.predictor._calculate_resource_contention(
            single_preds,
            "cuda",
            "parallel"
        )
        
        # Calculate sharing benefits
        benefits = self.predictor._calculate_sharing_benefits(
            self.model_configs,
            single_preds
        )
        
        # Calculate metrics
        metrics = self.predictor._calculate_multi_model_metrics(
            single_preds,
            contention,
            benefits,
            "parallel"
        )
        
        # Check metrics has expected keys
        self.assertIn("combined_throughput", metrics)
        self.assertIn("combined_latency", metrics)
        self.assertIn("combined_memory", metrics)
        
        # Check values are reasonable
        self.assertGreater(metrics["combined_throughput"], 0)
        self.assertGreater(metrics["combined_latency"], 0)
        self.assertGreater(metrics["combined_memory"], 0)
    
    def test_predict_multi_model_performance(self):
        """Test full multi-model performance prediction."""
        # Predict performance
        prediction = self.predictor.predict_multi_model_performance(
            self.model_configs,
            hardware_platform="cuda",
            execution_strategy="parallel"
        )
        
        # Check prediction has expected keys
        self.assertIn("total_metrics", prediction)
        self.assertIn("individual_predictions", prediction)
        self.assertIn("contention_factors", prediction)
        self.assertIn("sharing_benefits", prediction)
        self.assertIn("execution_schedule", prediction)
        
        # Check total metrics
        total_metrics = prediction["total_metrics"]
        self.assertIn("combined_throughput", total_metrics)
        self.assertIn("combined_latency", total_metrics)
        self.assertIn("combined_memory", total_metrics)
        
        # Check individual predictions
        self.assertEqual(len(prediction["individual_predictions"]), len(self.model_configs))
    
    def test_recommend_execution_strategy(self):
        """Test execution strategy recommendation."""
        # Get recommendation
        recommendation = self.predictor.recommend_execution_strategy(
            self.model_configs,
            hardware_platform="cuda",
            optimization_goal="throughput"
        )
        
        # Check recommendation has expected keys
        self.assertIn("recommended_strategy", recommendation)
        self.assertIn("optimization_goal", recommendation)
        self.assertIn("all_predictions", recommendation)
        
        # Check that a valid strategy was recommended
        self.assertIn(recommendation["recommended_strategy"], ["parallel", "sequential", "batched"])
        
        # Check that all strategies were evaluated
        self.assertEqual(len(recommendation["all_predictions"]), 3)
        
        # Check optimization goal
        self.assertEqual(recommendation["optimization_goal"], "throughput")
    
    def test_different_strategies(self):
        """Test prediction with different execution strategies."""
        # Test all strategies
        strategies = ["parallel", "sequential", "batched"]
        
        for strategy in strategies:
            # Predict performance
            prediction = self.predictor.predict_multi_model_performance(
                self.model_configs,
                hardware_platform="cuda",
                execution_strategy=strategy
            )
            
            # Check prediction is valid
            self.assertIn("total_metrics", prediction)
            self.assertIn("execution_schedule", prediction)
            
            # Check execution strategy is correct
            self.assertEqual(prediction["execution_strategy"], strategy)
            
            # Check schedule strategy matches
            self.assertEqual(prediction["execution_schedule"]["strategy"], strategy)
    
    def test_different_hardware(self):
        """Test prediction with different hardware platforms."""
        # Test multiple hardware platforms
        platforms = ["cpu", "cuda", "openvino", "webgpu"]
        
        for platform in platforms:
            # Predict performance
            prediction = self.predictor.predict_multi_model_performance(
                self.model_configs,
                hardware_platform=platform,
                execution_strategy="parallel"
            )
            
            # Check prediction is valid
            self.assertIn("total_metrics", prediction)
            self.assertIn("contention_factors", prediction)
            
            # Check hardware platform is correct
            self.assertEqual(prediction["hardware_platform"], platform)


if __name__ == "__main__":
    unittest.main()